external/easylogging++/easylogging++.cc

```diff
@@ -2968,6 +2968,16 @@ void Writer::initializeLogger(Logger *logger, bool needLock) {
 }
 
 void Writer::processDispatch() {
+  static std::atomic_flag in_dispatch;
```

Comment:
Is this to manage the case where an exception is thrown from the logger, and therefore its state needs to be reset? This is ... a hack. If something _must_ be cleaned up, like releasing a lock, that's the purpose for RAII in C++. Is the upstream easylogging++ implemented like this (improper lock cleanup)?

Reply:
Yes. In this case, an exception in sanitizing the string to be logged. I'd say yes to your last question, though it's not impossible they made sure no exception could arise. Sounds doubtful though.

Reply:
This chunk of code is done while a lock is held, so the `atomic_flag` can be a `bool` which will reduce the atomic instructions emitted. I don't think its possible for this chunk to execute without a lock - the entire function would likely have to be re-written.

Additionally, this likely fixes a particular symptom but not the cause. Calling the log function from the custom `__cxa_throw` is likely still problematic since the logger has multiple places where it can potentially throw. Luckily, these cases are rare. However, it seems prudent to disable stack printing on release builds entirely. Or at least make that the default. Our custom `__cxa_throw` calls several functions which can throw, and all of this occurs before the library starts it logic to track the number of active exceptions. Again, probably rare, but if easylogging++ throws on file write failures for instance, the result is likely a stackoverflow of attempted exception throws.

Reply:
This would defeat the purpose though, which is to get people to check whether interesting exceptions were thrown when something breaks.

Reply:
AFAICT only tools::stack_trace can throw, the rest is C. I suppose I could put an atomic flag in there to forbid reentrancy. That seems more encompassing...

Reply:
As a reminder - this `std::atomic_flag` can be a `bool`. This code is inside a lock.

> This would defeat the purpose though, which is to get people to check whether interesting exceptions were thrown when something breaks.

The information obtained doesn't worth the risk in hijacking the library routine. The daemon logs plenty already. A macro can achieve the same thing in critical areas without covertly hijacking control flow at _every_ `throw`.

> AFAICT only tools::stack_trace can throw, the rest is C. I suppose I could put an atomic flag in there to forbid reentrancy. That seems more encompassing...

Adding another atomic flag in the stack trace function is masking the symptom, just like this flag here. Easylogging++ was not written with the programmer thinking: "any potentially throwing function call can immediately call back into the top-level log routine" - the expectation is that easylogging++ functions will not do any logging. Hopefully the location for the flag is sufficient and not hastily selected.

And the "only one function can throw" is a misrepresentation - the easylogging++ log function or `std::cout` calls many C++ functions which can potentially throw. Most (probably nearly all) are `std::bad_alloc`, while others are from iostream. Either could result in infinite exceptions with the library being hijacked, but it should be uncommon. I did a quick scan (less than a few minutes) to find other likely to throw functions but nothing stood out (thankfully).